src/cholesky/lkjcholesky.jl

```diff
@@ -0,0 +1,183 @@
+        d > 0 || throw(ArgumentError("Matrix dimension must be positive."))
+        η > 0 || throw(ArgumentError("Shape parameter must be positive."))
+    end
+    logc0 = lkj_logc0(d, η)
+    uplo = _as_char(_uplo)
+    uplo ∈ ('U', 'L') || throw(ArgumentError("uplo must be 'U' or 'L'."))
+    T = Base.promote_eltype(η, logc0)
+    LKJCholesky{T, typeof(d)}(d, T(η), uplo, T(logc0))
+end
+
+_as_char(c::Char) = c
+_as_char(x) = only(string(x))
+
+#  -----------------------------------------------------------------------------
+#  REPL display
+#  -----------------------------------------------------------------------------
+
+show(io::IO, d::LKJCholesky) = show_multline(io, d, [(:d, d.d), (:η, d.η), (:uplo, d.uplo)])
+
+#  -----------------------------------------------------------------------------
+#  Conversion
+#  -----------------------------------------------------------------------------
+
+function convert(::Type{LKJCholesky{T}}, d::LKJCholesky) where T <: Real
+    return LKJCholesky{T, typeof(d.d)}(d.d, T(d.η), d.uplo, T(d.logc0))
+end
+function convert(::Type{LKJCholesky{T}}, d::Integer, η, logc0) where T <: Real
+    return LKJCholesky{T, typeof(d)}(d, T(η), d.uplo, T(logc0))
+end
+
+#  -----------------------------------------------------------------------------
+#  Properties
+#  -----------------------------------------------------------------------------
+
+dim(d::LKJCholesky) = d.d
+
+function size(d::LKJCholesky)
+    p = dim(d)
+    return (p, p)
+end
+
+function insupport(d::LKJCholesky, R::Cholesky)
+    p = dim(d)
+    factors = R.factors
+    (isreal(factors) && size(factors, 1) == p) || return false
+    iinds, jinds = axes(factors)
+    T = typeof(one(eltype(R))^2)
+    # check that the diagonal of U'*U or L*L' is all ones
+    @inbounds if R.uplo === 'U'
+        for j in 1:p
+            s = zero(T)
+            for i in 1:j
+                s += factors[iinds[i], jinds[j]]^2
+            end
+            isapprox(s, 1) || return false
+        end
+    else  # R.uplo === 'L'
+        for i in 1:p
+            s = zero(T)
+            for j in 1:i
+                s += factors[iinds[i], jinds[j]]^2
+            end
+            isapprox(s, 1) || return false
+        end
+    end
+    return true
+end
+
+function mode(d::LKJCholesky)
+    p = dim(d)
+    factors = Matrix{partype(d)}(I, p, p)
+    return Cholesky(factors, d.uplo, 0)
+end
+
+params(d::LKJCholesky) = (d.d, d.η, d.uplo)
+
+@inline partype(::LKJCholesky{T}) where {T <: Real} = T
+
+#  -----------------------------------------------------------------------------
+#  Evaluation
+#  -----------------------------------------------------------------------------
+
+function logkernel(d::LKJCholesky, R::Cholesky)
+    factors = R.factors
+    p, η = params(d)
+    c = p + 2(η - 1)
+    T = typeof(one(c) * log(one(eltype(factors))))
+    logp = zero(T)
+    di = diagind(factors)
+    for i in 2:p
+        logp += (c - i) * log(factors[di[i]])
+    end
```

Comment:
Again no need to use type heuristics, one can just compute the sum directly with `sum`.

Reply:
In general the trade-off we have is that summation via a loop is fastest but with greater code complexity and not as AD-friendly (e.g. with Zygote). e.g. here using `sum` is 10% slower than looping. I think this is acceptable.

Reply:
`sum` is slightly faster in the following simple benchmark:
```julia
julia> using BenchmarkTools

julia> using LinearAlgebra

julia> function f(c, x)
           T = typeof(one(c) * log(one(eltype(x))))
           di = diagind(x)               
           logp = zero(T)                
           for i in 2:length(di)         
               logp += (c - i) * log(x[di[i]])
           end  
           return logp   
       end             
f (generic function with 1 method)
          
julia> function g(c, x)
           logp = sum((c - i) * log(x[di]) for (i, di) in Iterators.drop(enumerate(diagind(x)), 1))
           return logp
       end      
g (generic function with 1 method)
                    
julia> c = rand();

julia> x = randn(40, 40);

julia> @benchmark f($c, $x)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     311.992 ns (0.00% GC)
  median time:      320.075 ns (0.00% GC)
  mean time:        339.895 ns (0.00% GC)
  maximum time:     740.527 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     241

julia> @benchmark g($c, $x)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     300.394 ns (0.00% GC)
  median time:      305.855 ns (0.00% GC)
  mean time:        321.193 ns (0.00% GC)
  maximum time:     690.703 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     249
```

Reply:
The speed advantage of `sum` goes away here of you already know the length of `di` (we do). In my `sum` implementations, it's always a little slower than the loops, but I've adopted it anyways.